### PR TITLE
Discard shader execution tests

### PR DESCRIPTION
**Pull request by alan-baker**
* Tests:
  * discard all
  * discard 3 quadrants
  * discard in function call
  * discard in various loops
  * derivatives and discard

Issue: #1683 

<hr>

**Requirements for PR author:**

- [X] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [X] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [X] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [X] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/execution/statement/discard.spec.ts [hunk 1]
```diff
@@ -0,0 +1,606 @@
...
+    const code = `
+${kSharedCode}
+
+fn foo(pos : vec2f) {
+  if (pos.x <= 0.5 && pos.y <= 0.5) {
+    discard;
+  }
+  if (pos.x >= 0.5 && pos.y >= 0.5) {
+    discard;
+  }
```
Comment (jrprice) [inline src/webgpu/shader/execution/statement/discard.spec.ts:322]: Again, I think these position checks are off, or the predicates aren't robust (or both). I can turn this into an unconditional discard and the test still passes.

**Review by jrprice — COMMENTED** [review]
Perhaps I'm missing something, but the shader logic and/or predicates seem off for a couple of tests.

File: src/webgpu/shader/execution/statement/discard.spec.ts [hunk 2]
```diff
@@ -0,0 +1,606 @@
...
+g.test('three_quarters')
+  .desc('Test a shader that discards all fragments')
+  .fn(t => {
+    const code = `
+${kSharedCode}
+
+@fragment
+fn fsMain(@builtin(position) pos : vec4f) -> @location(0) u32 {
+  _ = uniformValues[0];
+  if (pos.x >= 0.5 || pos.y >= 0.5) {
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply src/webgpu/shader/execution/statement/discard.spec.ts]: Whoops, thanks. I've updated the tests to use the right coordinates.

File: src/webgpu/shader/execution/statement/discard.spec.ts [hunk 3]
```diff
@@ -0,0 +1,617 @@
...
+  }
+  if pos.x >= 0.5 * ${kWidth} && pos.y >= 0.5 * ${kHeight} {
+    discard;
+  }
+}
+
+@fragment
+fn fsMain(@builtin(position) pos : vec4f) -> @location(0) u32 {
+  _ = uniformValues[0];
+  foo(pos.xy);
```
Comment (jrprice) [inline src/webgpu/shader/execution/statement/discard.spec.ts:337]: Likewise, the test still passes if this is changed to an unconditional discard.

**Review by jrprice — COMMENTED** [review]
The fragment positions LGTM now, but I'm still a little concerned about the fact that a couple of the tests don't fail if you make discards unconditional.

File: src/webgpu/shader/execution/statement/discard.spec.ts [hunk 4]
```diff
@@ -0,0 +1,617 @@
...
+    const dataChecker = (a: Float32Array) => {
+      return checkElementsPassPredicate(
+        a,
+        (idx: number, value: number | bigint) => {
+          const is_x = idx % 2 === 0;
+          if (is_x) {
+            return value < 0.5 * kWidth;
+          } else {
+            return value < 0.5 * kHeight;
+          }
```
[the remark this answers is not in the document]
Comment (alan-baker) [reply src/webgpu/shader/execution/statement/discard.spec.ts:275]: I've switched the render pass to load an initialized framebuffer with a sentinel larger than any possible write to detect these cases. I've re-tested 3/4 and function call with full discards and they now are caught as errors. Thanks for checking that.

File: src/webgpu/shader/execution/statement/discard.spec.ts [hunk 5]
```diff
@@ -212,7 +225,8 @@ fn fsMain(@builtin(position) pos : vec4f) -> @location(0) u32 {
       return checkElementsPassPredicate(
         a,
         (idx: number, value: number | bigint) => {
-          return value === 0;
+          return value === kWidth * kHeight;
+          //return value === 0;
```
Comment (jrprice) [inline src/webgpu/shader/execution/statement/discard.spec.ts:229]: Nit: commented code

**Review by jrprice — APPROVED** [review]
LGTM, thanks!

**Pull request merged**